lib/block_view_stake.go

```diff
@@ -0,0 +1,2600 @@
+	}
+
+	// Restore the PrevLockedStakeEntry, if exists. The PrevLockedStakeEntry will exist if the
+	// transactor has previously unstaked stake assigned to this validator within the same epoch.
+	// The PrevLockedStakeEntry will not exist otherwise.
+	// 1. Retrieve the CurrentLockedStakeEntry.
+	currentLockedStakeEntry, err := bav.GetLockedStakeEntry(
+		prevValidatorEntry.ValidatorPKID, transactorPKIDEntry.PKID, currentEpochNumber,
+	)
+	if err != nil {
+		return errors.Wrapf(err, "_disconnectUnstake: ")
+	}
+	// 2. Delete the CurrentLockedStakeEntry.
+	bav._deleteLockedStakeEntryMappings(currentLockedStakeEntry)
+	// 3. Set the PrevLockedStakeEntry, if exists.
+	if len(operationData.PrevLockedStakeEntries) > 1 {
+		return fmt.Errorf("_disconnectUnstake: more than one prev LockedStakeEntry found for %v", currentTxn.PublicKey)
+	}
+	if len(operationData.PrevLockedStakeEntries) == 1 {
+		bav._setLockedStakeEntryMappings(operationData.PrevLockedStakeEntries[0])
+	}
+
+	// Disconnect the basic transfer.
+	return bav._disconnectBasicTransfer(
+		currentTxn, txHash, utxoOpsForTxn[:operationIndex], blockHeight,
+	)
+}
+
+func (bav *UtxoView) _connectUnlockStake(
+	txn *MsgDeSoTxn,
+	txHash *BlockHash,
+	blockHeight uint32,
+	verifySignatures bool,
+) (
+	_totalInput uint64,
+	_totalOutput uint64,
+	_utxoOps []*UtxoOperation,
+	_err error,
+) {
+	// Validate the starting block height.
+	if blockHeight < bav.Params.ForkHeights.ProofOfStakeNewTxnTypesBlockHeight ||
+		blockHeight < bav.Params.ForkHeights.BalanceModelBlockHeight {
+		return 0, 0, nil, errors.Wrapf(RuleErrorProofofStakeTxnBeforeBlockHeight, "_connectUnlockStake: ")
+	}
+
+	// Validate the txn TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeUnlockStake {
+		return 0, 0, nil, fmt.Errorf(
+			"_connectUnlockStake: called with bad TxnType %s", txn.TxnMeta.GetTxnType().String(),
+		)
+	}
+
+	// Grab the txn metadata.
+	txMeta := txn.TxnMeta.(*UnlockStakeMetadata)
+
+	// Validate the txn metadata.
+	if err := bav.IsValidUnlockStakeMetadata(txn.PublicKey, txMeta); err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectUnlockStake: ")
+	}
+
+	// Convert TransactorPublicKey to TransactorPKID.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(txn.PublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return 0, 0, nil, errors.Wrapf(RuleErrorInvalidStakerPKID, "_connectUnlockStake: ")
+	}
+
+	// Convert ValidatorPublicKey to ValidatorPKID.
+	validatorPKIDEntry := bav.GetPKIDForPublicKey(txMeta.ValidatorPublicKey.ToBytes())
+	if validatorPKIDEntry == nil || validatorPKIDEntry.isDeleted {
+		return 0, 0, nil, errors.Wrapf(RuleErrorInvalidValidatorPKID, "_connectUnlockStake: ")
+	}
+
+	// Retrieve the PrevLockedStakeEntries. These will be restored if we disconnect this txn.
+	prevLockedStakeEntries, err := bav.GetLockedStakeEntriesInRange(
+		validatorPKIDEntry.PKID, transactorPKIDEntry.PKID, txMeta.StartEpochNumber, txMeta.EndEpochNumber,
```

Comment:
So... I might be missing something. But don't you need to verify that the prevLockedStakeEntries are actually allowed to be UNlocked? Right now, it looks like I can unstake and then immmediately unlock that stake right after, without a cooldown.

In addition, as I was reading this, it felt like UnlockStake should just unlock all the stake that is allowed to be unlocked. So you can get rid of StartEpochNumber and EndEpochNumber. Instead, GetLockedStakeENtriesInRange can just be GetUnlockableLockedStakeEntries, and it can search for LockedStakeEntries that have (LockedAtEpochNumber < CurrentEpochNumber+CooldownEpochs).

Reply:
Found it in the IsValidUnlockStake thing.